### PR TITLE
Add more details about cross-compilation

### LINKED ISSUE #1534: Introduction suggests no Apple ARM packages
- [X] I read [the conda-forge documentation](https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge) and could not find the solution for my problem there.

As an Apple user I am interested in the status of conda and conda-forge on the new Apple silicon ARM machines like the M1 CPU.

Quoting https://conda-forge.org/docs/user/introduction.html

> *By default, we build packages for macOS, Linux AMD64 and Windows AMD64.*

This suggests osx-arm64 is not supported, where as https://conda-forge.org/blog/posts/2020-10-29-macos-arm64/ suggests this is quite mature now.

Suggested wording:

> *By default, we build packages for macOS (AMD64 and ARM64), Linux AMD64 and Windows AMD64.*

### LINKED ISSUE #1841: Improve documentation on using emulation & cross-compilation
While there are a few mentions of emulation & cross-compilation in various points in the docs (like [Configuring `conda-forge.yml`]( https://conda-forge.org/docs/maintainer/conda_forge_yml.html ) or [Apple Silicon builds]( https://conda-forge.org/docs/maintainer/knowledge_base.html#apple-silicon-builds )), there doesn't appear to be (or at least basic searching doesn't uncover it) a step-by-step walkthrough of what things a maintainer needs to do to setup an emulated build or enable cross-compilation. Would be good to have this to help more users get started on their own with these options (particularly when native builds fall over).

### LINKED ISSUE #1983: osx_arm64: more detailed instructions
### Where should the content be added?

https://conda-forge.org/docs/maintainer/knowledge_base.html

### What should be added?

I could not find detailed instructions for adding osx-arm64 support for an MPI package and succeeded only with the help of @isuruf. Some information is available on
- https://conda-forge.org/docs/maintainer/knowledge_base.html#cross-compilation
- https://conda-forge.org/blog/posts/2020-10-29-macos-arm64

An related issue is https://github.com/conda-forge/conda-forge.github.io/issues/1534, which also mentions a migrator (https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/migrations/osx_arm64.txt).

There are two routes for an improved documentation: Either have add an extra section on (cross)-compiling for osx-arm64 or add content to https://conda-forge.org/docs/maintainer/knowledge_base.html#cross-compilation and https://conda-forge.org/docs/maintainer/knowledge_base.html#message-passing-interface-mpi. This should be consistent with the approach for other non-default architectures.

### Additional information

Content of the documentation
- https://conda-forge.org/blog/posts/2020-10-29-macos-arm64 is a good starting point, most of it should become part of the official documentation to emphasize that this is the current state and not just an potentially outdated blog post.
- for https://github.com/conda-forge/libfyaml-feedstock (autotools) `cp $BUILD_PREFIX/share/gnuconfig/config.* .` was not required.
- build requirement for mpi: ` openmpi  # [mpi == 'openmpi' and build_platform != target_platform]` (with explicit explanation why openMPI and MPICH behave differently).

### PR DESCRIPTION
**Pull request by jaimergp**
PR Checklist:

- [X] note any issues closed by this PR with [closing keywords](https://help.github.com/articles/closing-issues-using-keywords)
- [X] put any other relevant information below

Closes #1983 by consolidating the information available in the [osx-arm64 blog post](https://conda-forge.org/blog/posts/2020-10-29-macos-arm64/) written by @isuruf. 

Will also:
- close #1534
- close #1841

### CODE REVIEW

**Comment by jakirkham** [comment]
Thanks Jaime! 🙏

Wonder if we could address this issue ( https://github.com/conda-forge/conda-forge.github.io/issues/1841 )

cc @ahesford @stefanv

File: src/maintainer/knowledge_base.rst [hunk 1]
```diff
@@ -303,14 +303,68 @@ A package that needs all five compilers would define
...
+  the build scripts (e.g. ``$build_platform``), and also as Jinja variables in the ``meta.yaml``
+  selectors (e.g. ``# [build_platform != target_platform]``).
+
+In addition to these two variables, there are some more environment variables that are set by
+conda-forge's automation (e.g. ``conda-forge-ci-setup``, compiler activation packages, etc) that
+can aid in cross-compilation setups:
+
+- ``CONDA_BUILD_CROSS_COMPILATION``: set to ``1`` when ``build_platform`` and ``target_platform``
+  differ.
+- ``BUILD_PLATFORM``: the platform on which ``conda-build`` is running.
```
Comment (isuruf) [inline src/maintainer/knowledge_base.rst:343]:
```suggestion
```

File: src/maintainer/knowledge_base.rst [hunk 2]
```diff
@@ -303,14 +303,68 @@ A package that needs all five compilers would define
...
+- ``BUILD``: the autoconf triplet expected for build platform.
+- ``HOST_PLATFORM``: the platform on which the package will be installed.
```
Comment (isuruf) [inline src/maintainer/knowledge_base.rst:345]:
```suggestion
```

File: src/maintainer/knowledge_base.rst [hunk 3]
```diff
@@ -303,14 +303,68 @@ A package that needs all five compilers would define
...
+  selectors (e.g. ``# [build_platform != target_platform]``).
+
+In addition to these two variables, there are some more environment variables that are set by
+conda-forge's automation (e.g. ``conda-forge-ci-setup``, compiler activation packages, etc) that
+can aid in cross-compilation setups:
+
+- ``CONDA_BUILD_CROSS_COMPILATION``: set to ``1`` when ``build_platform`` and ``target_platform``
+  differ.
+- ``BUILD_PLATFORM``: the platform on which ``conda-build`` is running.
+- ``BUILD``: the autoconf triplet expected for build platform.
```
Comment (isuruf) [inline src/maintainer/knowledge_base.rst:344]:
```suggestion
- ``CONDA_TOOLCHAIN_BUILD``: the autoconf triplet expected for build platform.
```

File: src/maintainer/knowledge_base.rst [hunk 4]
```diff
@@ -303,14 +303,68 @@ A package that needs all five compilers would define
...
+- ``HOST_PLATFORM``: the platform on which the package will be installed.
+- ``HOST``: the autoconf triplet expected for host platform.
```
Comment (isuruf) [inline src/maintainer/knowledge_base.rst:346]:
```suggestion
- ``CONDA_TOOLCHAIN_HOST``: the autoconf triplet expected for host platform.
```

File: src/maintainer/knowledge_base.rst [hunk 5]
```diff
@@ -335,6 +388,46 @@ In the build script, it would need to update the config files and guard any test
...
+
+In the build script, it would need to update ``cmake`` call and guard any tests when cross-compiling:
+
+.. code-block:: sh
+
+    # Pass ``CMAKE_ARGS`` to ``cmake``
+    cmake ${CMAKE_ARGS} ..
+    
+    # Skip ``ctest`` when cross-compiling
+    if [[ "${CONDA_BUILD_CROSS_COMPILATION}" != "1" ]]; then
```
Comment (isuruf) [inline src/maintainer/knowledge_base.rst:409]:
```suggestion
    if [[ "${CONDA_BUILD_CROSS_COMPILATION:-}" != "1" || "${CROSSCOMPILING_EMULATOR:-}" != "" ]]; then
```

**Comment by jaimergp** [comment]
Added a couple sections more based on that (excellent) reference @jakirkham.

**Comment by jaimergp** [comment]
Thanks for your suggestions @isuruf! Btw, what's `$CROSSCOMPILING_EMULATOR` and where is it being set? Do we need to document it next to the other env vars?

**Comment by isuruf** [comment]
It's set to the QEMU executable. See https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/main/recipe/cross_compile_support.sh#L36-L37

File: src/maintainer/knowledge_base.rst [hunk 6]
```diff
@@ -486,9 +528,58 @@ following changes before the builds scripts run:
...
+    provider:
+      linux_aarch64: travis
+      linux_ppc64le: azure
+      linux_64: azure
+
+Use this variables in your builds scripts if needed:
+
+- ``CONDA_BUILD_CROSS_COMPILATION``: Set to ``1`` when build and host platforms differ. This can
+  mean your are cross-compiling or emulating.
+- ``CROSSCOMPILING_EMULATOR``: Path to the ``qemu`` binary for the host platform.
```
Comment (isuruf) [inline src/maintainer/knowledge_base.rst:579]: This emulator is for cross compiling while the text around it has been for native compilation (full OS emulation is still native compilation).
This emulator is used when cross compiling and is used for running the tests after compiling is done. Or for when running some executables by the build system.

Comment (jaimergp) [reply src/maintainer/knowledge_base.rst:579]: Ohh I see, ok ok. I'll remove it from here then. Thanks!

**Review by xhochy — APPROVED** [review]

**Comment by xhochy** [comment]
This needs a rebase.

**Comment by jaimergp** [comment]
Rebased! Preview available at https://deploy-preview-1985--conda-forge-previews.netlify.app/docs/maintainer/knowledge_base/#cross-compilation.

**Review by xhochy — APPROVED** [review]

**Pull request merged**